packages/apps/dashboard/ui-2024/src/features/Leaderboard/hooks/useDataGrid.tsx

```diff
@@ -0,0 +1,190 @@
+import { Box, Typography } from '@mui/material';
+import { GridColDef, GridRenderCellParams } from '@mui/x-data-grid';
+import { LeaderBoardData } from '@services/api/use-leaderboard-details';
+import { useBreakPoints } from '@utils/hooks/use-is-mobile';
+import { useLeaderboardSearch } from '@utils/hooks/use-leaderboard-search';
+import HelpOutlineIcon from '@mui/icons-material/HelpOutline';
+import CustomTooltip from '@components/CustomTooltip';
+import { colorPalette } from '@assets/styles/color-palette';
+import { useMemo } from 'react';
+
+import { RoleCell } from '../components/RoleCell';
+import { AddressCell } from '../components/AddressCell';
+import { ChainCell } from '../components/ChainCell';
+import { SelectNetwork } from '../components/SelectNetwork';
+import { ReputationLabel } from '../components/ReputationLabel';
+import { TextCell } from '../components/TextCell';
+
+export const useDataGrid = (data: LeaderBoardData) => {
+  const {
+    filterParams: { chainId },
+  } = useLeaderboardSearch();
+
+  const {
+    mobile: { isMobile },
+  } = useBreakPoints();
+  const formattedData = useMemo(() => {
+    return data.map((row, index) => {
+      return {
+        ...row,
+        id: index,
```

Comment:
AFAIK data-grid relies on `id` for different internal things, so I would not use it for row index to avoid implicit/unexpected issues.

Would be better to use `address` as `id` and have `rowIndex: index`

Reply:
good point, thanks! I just realized that  we don't need an index for a row because we are checking indexOf a row by id so I changed this id from index to address.
![Zrzut ekranu 2024-11-21 o 09 54 15](https://github.com/user-attachments/assets/6b49b7b0-9522-45a1-926b-4b45f6150999)

Reply:
So here we can update just a row
![Zrzut ekranu 2024-11-21 o 09 58 51](https://github.com/user-attachments/assets/b3225309-df06-42dc-9dfb-5d9f615c674e)

Reply:
> I just realized that we don't need an index for a row because we are checking indexOf a row

And that was exactly what we were trying to avoid :)

This `sortedRows.indexOf` won't return the proper value until the whole grid is re-rendered (and you achieve it implicitly by calling `updateRow([row])`. The latter in `handleSortModelChange` looks strange: why do we have to update the row with same data on sort model change if `getSortedRows` already has it?! That thing with row index is still "implicit magic" and hard to catch for someone how is not deep into data-grid behavior. Instead - code should be explicit so it's easy to maintain.

This is why I'm suggesting to have `rowIndex` property for data, update it in `handleSortModelChange` and use it as separate column

Reply:
I investigated it a bit and this update from `handleSortModelChange` does not cause the component to be rerendered. So I can't make it working with rowIndex field. I fully agree this solution is more readable and easier to maintain but at the moment I have no idea how to implement it differently.

I can get correct order by refering this state `params.api.state.sorting.sortedRows`

I have 2 console.logs:
1. refers to `params.api.state.sorting.sortedRows`
2. refers to updated `rowIndex` which remains same after sorting what you can check in second screenshot

![Zrzut ekranu 2024-11-21 o 11 25 17](https://github.com/user-attachments/assets/24c42ccb-df3c-471a-8434-5330de804886)

![Zrzut ekranu 2024-11-21 o 11 23 19](https://github.com/user-attachments/assets/0c283585-daed-4d72-bcdd-bb54c3205795)

It turns out that this `apiRef.current.updateRows([{ ...row, rowIndex: idx }])` will not rerender component

Reply:
@adrian-oleskiewicz 

Tricky part is that `details.api.getSortedRows();` doesn't give you rows sorted by new sort model, so the order of operations is next:
1. you click on column header
2. sort model is updated in data grid
3. `handleSortModelChange` callback invoked
4. data sorted in data grid

You can easily verify this by adding logs to `handleSortModelChange` right 
```
const sortedRows = details.api.getSortedRows();
console.log(
  'sorted',
  sortedRows.map((r, index) => ({
    id: r.id,
    old: r.rowIndex,
    new: index,
  }))
);
setTimeout(() => {
  const sortedRows = details.api.getSortedRows();
  console.log(
    'sorted 2',
    sortedRows.map((r, index) => ({
      id: r.id,
      old: r.rowIndex,
      new: index,
    }))
  );
}, 1000);
```
I don't know what is their internal implementation (probably they rely on references and memo some stuff), but when the data is sorted - the do not call `renderCell`, even though the DOM is updated. In case you call `updateRows` - it causes `renderCell` call.

So, taking the above details into account, we can actually make it working using `rowIndex` because luckily there is ["applySorting"](https://mui.com/x/api/data-grid/grid-api/#grid-api-prop-applySorting) util in its API, i.e. `handleSortModelChange` should be
```
apiRef.current.applySorting();
// now rows are actually sorted
const sortedRows = details.api.getSortedRows();
sortedRows.forEach((row, index) => {
  apiRef.current.updateRows([{ ...row, rowIndex: index }]);
});
```

Seems like together we finally got how to work with it properly 😅

Reply:
@dnechay Great job! You found the missing puzzle. Thanks for such a good code review!